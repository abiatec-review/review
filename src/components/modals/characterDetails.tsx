import TextRow from '@components/textRow';
import useOrientation from '@hooks/useOrientation';
import Character from '@models/character';
import React from 'react';
import {Image, Platform, StyleSheet, View} from 'react-native';

import Modal from '.';

interface Props {
  character: Character;
  isShown: boolean;
  toggle: () => void;
}

function CharacterDetailsModal(props: Props) {
  const {character, isShown, toggle} = props;

  const {isPortrait} = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapeStyles;

  return (
    <Modal isShown={isShown} toggle={toggle} style={styles.modal}>
      <View style={styles.infoBlock}>
        <Image source={{uri: character.image}} style={styles.image} />
        <View style={styles.textBlock}>
          <TextRow field="Status" data={character.status} />
          <TextRow field="Gender" data={character.gender} />
          <TextRow field="Origin" data={character.origin?.name} />
          <TextRow field="Created" data={character.created} />
        </View>
      </View>
    </Modal>
  );
}

export default CharacterDetailsModal;

const baseStyles = StyleSheet.create({
  image: {
    aspectRatio: 1,
    borderRadius: 20,
  },
  textBlock: {
    flex: 1,
    justifyContent: 'space-evenly',
  },
});

const portraitStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: Platform.select({ios: 0.65, android: 0.85}),
  },
  infoBlock: {flex: 1},
  image: {
    ...baseStyles.image,
    width: '100%',
  },
});

const landscapeStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: 0.62,
  },
  infoBlock: {
    marginBottom: 10,
    alignItems: 'center',
    flexDirection: 'row',
  },
  image: {
    ...baseStyles.image,
    flex: Platform.select({ios: 0.45, android: 0.5}),
    marginRight: 15,
  },
  textBlock: {
    ...baseStyles.textBlock,
    height: "80%"
  }
});
