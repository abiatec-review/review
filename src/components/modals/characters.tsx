import useOrientation from '@hooks/useOrientation';
import Character from '@models/character';
import React from 'react';
import * as RN from 'react-native';

import Modal from '.';

interface Props {
  title: string;
  isShown: boolean;
  toggle: () => void;
  characters: Array<Character>;
}

function CharactersModal(props: Props) {
  const {characters, isShown, toggle, title} = props;

  const {isPortrait} = useOrientation();

  const getHeight = () => {
    if (characters.length > 10) {
      return {height: isPortrait ? 270 : 180};
    } else {
      const multiplier = characters.length % 10;
      if (isPortrait) return {height: multiplier > 3 ? 270 : multiplier * 90};
      else return {height: multiplier > 3 ? 180 : multiplier * 90};
    }
  };

  return (
    <Modal isShown={isShown} toggle={toggle} style={{flex: 0}}>
      <RN.Text style={styles.title}>{title}</RN.Text>
      <RN.ScrollView style={getHeight()}>
        {characters.map(({id, name, image}) => (
          <RN.View key={id} style={styles.character}>
            <RN.Image source={{uri: image}} style={styles.image} />
            <RN.Text style={styles.text}>{name}</RN.Text>
          </RN.View>
        ))}
      </RN.ScrollView>
    </Modal>
  );
}

export default CharactersModal;

const styles = RN.StyleSheet.create({
  title: {
    fontSize: 20,
    paddingBottom: 10,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  character: {
    marginBottom: 10,
    alignItems: 'center',
    flexDirection: 'row',
  },
  image: {
    width: 80,
    height: 80,
    marginRight: 10,
    borderRadius: 50,
  },
  text: {
    flex: 1,
    fontSize: 18,
    fontWeight: '600',
  },
});
