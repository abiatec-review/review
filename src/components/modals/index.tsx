import useOrientation from '@hooks/useOrientation';
import React from 'react';
import * as RN from 'react-native';

interface Props {
  isShown: boolean;
  toggle: () => void;
  style?: RN.FlexStyle;
}

function Modal(props: React.PropsWithChildren<Props>) {
  const {isShown, toggle, children, style} = props;

  const {isPortrait} = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapelStyles;

  return (
    <RN.Modal
      transparent
      animationType="fade"
      visible={isShown}
      onRequestClose={toggle}
      supportedOrientations={['portrait', 'landscape']}>
      <RN.View style={styles.container}>
        <RN.View style={[styles.content, style]}>
          {children}
          <RN.Pressable style={styles.closeButton} onPress={toggle}>
            <RN.Text style={styles.closeText}>Close</RN.Text>
          </RN.Pressable>
        </RN.View>
      </RN.View>
    </RN.Modal>
  );
}

export default Modal;

const baseStyles = RN.StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  content: {
    padding: 20,
    borderRadius: 20,
    backgroundColor: 'white',
  },
  closeButton: {
    width: '100%',
    padding: 10,
    borderRadius: 20,
    backgroundColor: '#2196F3',
  },
  closeText: {
    fontSize: 18,
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

const portraitStyles = RN.StyleSheet.create({
  ...baseStyles,
  content: {
    ...baseStyles.content,
    width: '91%',
  },
});

const landscapelStyles = RN.StyleSheet.create({
  ...baseStyles,
  content: {
    ...baseStyles.content,
    width: '70%',
  },
});
