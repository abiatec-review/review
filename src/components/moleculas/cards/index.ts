export { FullCharacterCard } from "./fullCharacter";
export { ReducedCharacterCard } from "./reducedCharacter";
export { LocationCard } from "./location";
export { EpisodeCard } from "./episode";
