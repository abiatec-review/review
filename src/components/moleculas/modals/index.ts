export { CharactersModal } from "./characters";
export { CharacterDetailsModal } from "./characterDetails";
