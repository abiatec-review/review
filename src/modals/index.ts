export { CharacterFullInfo } from './CharacterFullInfo';
export { ImagePickerModal } from './ImagePickerModal';
