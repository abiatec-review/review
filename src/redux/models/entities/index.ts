export { Character } from "./character";
export { Location } from "./location";
export { Episode } from "./episode";
export { Pagination } from "./pagination";
export { ResultList } from "./pagination";
