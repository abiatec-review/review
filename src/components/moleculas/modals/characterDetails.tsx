import React from "react";

import { Image, Platform, StyleSheet, Text, View } from "react-native";

import { TextRow } from "@components/atoms";
import { Modal } from "@components/atoms";
import { useOrientation } from "@hooks";
import { Character } from "@redux/models/entities";
import { FontSize, Indent, Radius } from "@utils";

interface Props {
  character: Character;
  isShown: boolean;
  toggle: () => void;
}

export function CharacterDetailsModal(props: Props) {
  const { character, isShown, toggle } = props;

  const { isPortrait } = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapeStyles;

  return (
    <Modal isShown={isShown} toggle={toggle} style={styles.modal}>
      <Text style={styles.name}>{character.name}</Text>
      <View style={styles.infoBlock}>
        <Image source={{ uri: character.image }} style={styles.image} />
        <View style={styles.textBlock}>
          <TextRow field="Status" data={character.status} />
          <TextRow field="Gender" data={character.gender} />
          <TextRow field="Origin" data={character.origin?.name} />
          <TextRow field="Created" data={character.created} />
        </View>
      </View>
    </Modal>
  );
}

const baseStyles = StyleSheet.create({
  name: {
    fontWeight: "bold",
    textAlign: "center",
    fontSize: FontSize.MEDIUM
  },
  image: {
    aspectRatio: 1,
    borderRadius: Radius.DEFAULT
  },
  infoBlock: {
    flex: 1
  },
  textBlock: {
    flex: 0.9,
    justifyContent: "space-evenly"
  }
});

const portraitStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: Platform.select({ ios: 0.65, android: 0.85 })
  },
  name: {
    ...baseStyles.name,
    marginBottom: Indent.DEFAULT
  },
  image: {
    ...baseStyles.image,
    width: "100%"
  }
});

const landscapeStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: 0.6
  },
  infoBlock: {
    ...baseStyles.infoBlock,
    flexDirection: "row",
    alignItems: "center"
  },
  image: {
    ...baseStyles.image,
    marginRight: Indent.HUGE,
    flex: 0.5
  },
  textBlock: {
    ...baseStyles.textBlock,
    flex: 1,
    height: "80%"
  }
});
