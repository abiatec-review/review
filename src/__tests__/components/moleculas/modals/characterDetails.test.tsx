import React from "react";

import { create } from "react-test-renderer";

import { CharacterDetailsModal } from "@components/moleculas/modals";
import { Character } from "@redux/models/entities";

jest.mock("@react-native-firebase/auth", () => {});
jest.mock("@react-native-firebase/storage", () => {});
jest.mock("react-native-image-picker", () => {});
jest.mock("react-native-vector-icons/MaterialIcons", () => {});

let mockOrientation = true;
jest.mock("@hooks", () => ({ useOrientation: () => ({ isPortrait: mockOrientation }) }));

const character: Character = {
  id: 1,
  name: "test",
  type: "test",
  gender: "test",
  status: "test",
  species: "test",
  origin: {
    name: "test",
    url: "test"
  },
  location: {
    name: "test",
    url: "test"
  },
  image: "test",
  episode: ["test"],
  url: "test",
  created: "test"
};

describe("CharacterDetailsModal render", () => {
  const modalProps = { isShown: true, toggle: () => {}, character };

  test("portrait", async () => {
    const tree = create(<CharacterDetailsModal {...modalProps} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test("landscape", () => {
    mockOrientation = false;
    const tree = create(<CharacterDetailsModal {...modalProps} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
