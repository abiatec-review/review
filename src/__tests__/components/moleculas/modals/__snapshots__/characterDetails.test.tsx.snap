// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailsModal render landscape 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "rgb(175, 224, 209)",
            "borderRadius": 20,
            "paddingHorizontal": 20,
            "paddingVertical": 10,
            "width": "70%",
          },
          Object {
            "flex": 0.6,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        test
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "test",
            }
          }
          style={
            Object {
              "aspectRatio": 1,
              "borderRadius": 20,
              "flex": 0.5,
              "marginRight": 15,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "height": 170,
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Status:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Gender:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Origin:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Created:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`CharacterDetailsModal render portrait 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "rgb(175, 224, 209)",
            "borderRadius": 20,
            "paddingHorizontal": 20,
            "paddingVertical": 10,
            "width": "91%",
          },
          Object {
            "height": "68%",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginBottom": 10,
            "textAlign": "center",
          }
        }
      >
        test
      </Text>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "test",
            }
          }
          style={
            Object {
              "aspectRatio": 1,
              "borderRadius": 20,
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "height": 170,
              "justifyContent": "space-between",
              "marginTop": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Status:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Gender:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Origin:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "black",
                  "fontSize": 18,
                  "marginRight": 20,
                }
              }
            >
              Created:
            </Text>
            <Text
              style={
                Object {
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              test
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
